### ISSUE TITLE
[Feature request] Android Automotive support

### ISSUE THREAD

**Issue by chhjel** [osmandapp/OsmAnd#15981]
# 🚀 feature request

### Description

Support for Android Automotive and not just Android Auto.

### Describe the solution you'd like

Would be nice with a version for Android Automotive so that a phone would not be required to use the app.

### Describe alternatives you've considered

No good alternatives as I'm aware.

**Comment by lbdroid** [comment]
It does work in Android Automotive, but you have to launch it from the drawer, which means it will only work in full screen mode.

I presume what you're asking is to make it advertise itself as a navigation application for Android Automotive.

This can be accomplished by adding an intent filter category to the manifest;
```<category android:name="android.intent.category.APP_MAPS"/>```
https://android.googlesource.com/platform/packages/services/Car/+/refs/heads/master/car-maps-placeholder/AndroidManifest.xml#35

**Comment by chhjel** [comment]
Hmm can't seem to find the app at all on the app store from a Polestar 2, I'll do a better check in a few days and see if I can make it work 😃

**Comment by lbdroid** [comment]
I don't know what "app store" you are referring to, but it is available through F-Droid. If you're referring to something like the gooble mess, it filters availability by flags, so it would have to be advertised there as for cars. Ignore that and install directly, like with ADB or something.

**Comment by chhjel** [comment]
The app store is the Google Play store, but the app selection is quite limited on Android Automotive OS. There's not that many apps currently available currently:
https://play.google.com/store/apps/collection/promotion_android_auto_embedded_exploreall 

Developer Mode is blocked in Android Automotive OS, there's no USB-Debugging available, and installing not-trusted apks is also disabled, at least not on the Polestar 2 as I'm aware.

It seems the only way other people are installing apps from outside of the app-store is by building the apps themselves, publishing the app using internal beta testing, and installing it from a web browser instead of the app store in the car. (https://forum.xda-developers.com/t/help-sideload-apps-polestar-2.4383145/)

Seems the app must go through some sort of review for it to be published on Android Automotive (https://developer.android.com/training/cars/distribute), I was hoping that could be done for OsmAnd so that the app would be visible on the app store within the car 😃

**Comment by lbdroid** [comment]
> installing it from a web browser instead of the app store in the car.

@chhjel ; Its pretty terrible that they've crippled the build you're stuck with so thoroughly. IMO, that should be a crime.
Regarding installation through a web browser; there shouldn't be any need to build the application yourself, just make it available through a web server from some location you can retrieve it. That could even be something like a git repository! You have an account on github, so create a repository and add your apk's to it, then install from the web browser in the car.

Yes, I know its a hack, but hopefully useful for you between now and when this "feature" is implemented.

**Comment by landswellsong** [comment]
+1 here
There is another specific EV use-case, navigation to a charger can activate the battery precondition routine. The native google maps app does that so there might be some kind of API for doing it in OsmAnd.

**Comment by TheOneAndOnlyPP** [comment]
I really hope that Osmand can add support for Android Automotive OS (AAOS). More and more car models is launching with AA and there is a growing demand for map services that is an alternative to Google Maps. In the Wikipedia article below, they list 90+ models that is currently known, both with and without Google Services. 

https://en.m.wikipedia.org/wiki/Android_Automotive

The comment from @chhjel ahows exactly what is needed. To use the built in feature of the Polestar 2 to precondition the battery 30 minutes before you reach the charger that the navigation is set to.

**Comment by yekibud** [comment]
> Seems the app must go through some sort of review for it to be published on Android Automotive (https://developer.android.com/training/cars/distribute), I was hoping that could be done for OsmAnd so that the app would be visible on the app store within the car 😃

OSMAnd devs:  It's been two years since this post and still sounds like no OSMAnd available on Android Automotive?  Isn't it just a matter of submitting the app for publication as @chhjel suggested?  What are the obstacles that need to be overcome?  Is there any dev work?

**Comment by ftab** [comment]
> This can be accomplished by adding an intent filter category to the manifest;
> `<category android:name="android.intent.category.APP_MAPS"/>`
> https://android.googlesource.com/platform/packages/services/Car/+/refs/heads/master/car-maps-placeholder/AndroidManifest.xml#35

This doesn't seem to work, unless I'm failing to find the right spot to put it in. I tried this but it doesn't offer to become the system's maps app.

```diff
diff --git a/OsmAnd/AndroidManifest.xml b/OsmAnd/AndroidManifest.xml
index 13c551e7d8..a0cad79deb 100644
--- a/OsmAnd/AndroidManifest.xml
+++ b/OsmAnd/AndroidManifest.xml
@@ -127,6 +127,7 @@
                                <action android:name="android.intent.action.VIEW" />
                                <action android:name="android.intent.action.SEND" />
                                <category android:name="android.intent.category.DEFAULT" />
+                <category android:name="android.intent.category.APP_MAPS"/>
                                <data android:scheme="content" android:host="*" android:mimeType="binary/octet-stream" />
                                <data android:scheme="content" android:host="*" android:mimeType="application/octet-stream" />
                        </intent-filter>
@@ -135,6 +136,7 @@
                                <action android:name="android.intent.action.MAIN" />
                                <category android:name="android.intent.category.LAUNCHER" />
                                <category android:name="android.intent.category.MULTIWINDOW_LAUNCHER" />
+                <category android:name="android.intent.category.APP_MAPS"/>
                        </intent-filter>
 
                        <!-- Open osmand.net/go links in OsmAnd app by default (without showing chooser dialog). -->
```

It's also not using the Android Auto style UI. Can anyone show me how to enable that?

**Comment by afkdev04** [comment]
> > This can be accomplished by adding an intent filter category to the manifest;
> > `<category android:name="android.intent.category.APP_MAPS"/>`
> > https://android.googlesource.com/platform/packages/services/Car/+/refs/heads/master/car-maps-placeholder/AndroidManifest.xml#35
> 
> This doesn't seem to work, unless I'm failing to find the right spot to put it in. I tried this but it doesn't offer to become the system's maps app.
> 
> diff --git a/OsmAnd/AndroidManifest.xml b/OsmAnd/AndroidManifest.xml
> index 13c551e7d8..a0cad79deb 100644
> --- a/OsmAnd/AndroidManifest.xml
> +++ b/OsmAnd/AndroidManifest.xml
> @@ -127,6 +127,7 @@
>                                 <action android:name="android.intent.action.VIEW" />
>                                 <action android:name="android.intent.action.SEND" />
>                                 <category android:name="android.intent.category.DEFAULT" />
> +                <category android:name="android.intent.category.APP_MAPS"/>
>                                 <data android:scheme="content" android:host="*" android:mimeType="binary/octet-stream" />
>                                 <data android:scheme="content" android:host="*" android:mimeType="application/octet-stream" />
>                         </intent-filter>
> @@ -135,6 +136,7 @@
>                                 <action android:name="android.intent.action.MAIN" />
>                                 <category android:name="android.intent.category.LAUNCHER" />
>                                 <category android:name="android.intent.category.MULTIWINDOW_LAUNCHER" />
> +                <category android:name="android.intent.category.APP_MAPS"/>
>                         </intent-filter>
>  
>                         
> 
> It's also not using the Android Auto style UI. Can anyone show me how to enable that?

edit: sry first time using the github reply formatting
still testing but i got it to work (havent used it very much but it seems to be complete might want to modify the ui a little bit for the smaller window tho) also its only worked in the android studio emulator (the old automotive 14 os) i have aosp for automotive os 16 built but i havent gotten this to work in that emulator yet

remove the entirety of those two intent filters (the first two sections under net.osmand.plus.activities.MapActivity)
`	<intent-filter>
 		<action android:name="android.intent.action.VIEW" />
		<action android:name="android.intent.action.SEND" />
		<category android:name="android.intent.category.DEFAULT" />
		<category android:name="android.intent.category.APP_MAPS"/>
		<data android:scheme="content" android:host="*" android:mimeType="binary/octet-stream" />
		<data android:scheme="content" android:host="*" android:mimeType="application/octet-stream" />
	</intent-filter>
	<intent-filter>
		<action android:name="android.intent.action.MAIN" />
		<category android:name="android.intent.category.LAUNCHER" />
		<category android:name="android.intent.category.MULTIWINDOW_LAUNCHER" />
		<category android:name="android.intent.category.APP_MAPS"/>
	</intent-filter>`


and replace with just this one
`	<intent-filter>
		<action android:name="android.intent.action.MAIN" />
		<category android:name="android.intent.category.DEFAULT" />
		<category android:name="android.intent.category.LAUNCHER" />
		<category android:name="android.intent.category.APP_MAPS" />
	</intent-filter>`

**Comment by ftab** [comment]
@conradkleinpeter nice work! That gets it into the home screen. Thank you! At long last, I have something other than the big ugly green "No maps application installed" screen staring at me there.

Now I'm looking for some way to activate the driving-optimized UI from the Android Auto version... 🤔 those small tablet sized text and buttons are a little tricky to read and hit.

edited to add screenshot

<img width="1280" height="800" alt="Image" src="https://github.com/user-attachments/assets/17d3bbcc-5782-4ec7-8ad1-16c87c180b09" />

**Comment by landswellsong** [comment]
Is there a way this can be built and pushed as some "beta testing" channel or so on to try on the cars that disallow sideloading? Or should I volounteer?

**Comment by ftab** [comment]
I've made a fork of OsmAnd to get the Android Auto UI to display on Android Automotive

<img width="1280" height="800" alt="Image" src="https://github.com/user-attachments/assets/72487f5e-04c2-4a6d-8bfb-8967df2a3ad8" />

https://github.com/radiosound-com/OsmAnd/tree/caramel-vanilla-osmand-aaos

Compare: https://github.com/osmandapp/OsmAnd/compare/master...radiosound-com:OsmAnd:caramel-vanilla-osmand-aaos

TL;DR: Add an automotive build flavor & manifest: CarAppActivity for driving, separate full UI in the app launcher for parked tasks (map download and other things that aren't in the driving optimized ui, etc). I started to modify the original manifest but then realized it was going to wreak havoc with running it on regular Android...

Requesting developer / maintainer review to help get this mergeable. Do I just go ahead and open the PR? I'm not sure if it's all the way ready yet or if this approach is even acceptable...

**Comment by ftab** [comment]
@vshcherb hope you don't mind a gentle ping, would you have time to check out the above or know someone who can help? I've got OsmAnd working in Android Automotive but it's a new build target in this case as the manifest is significantly different than the regular one, to provide for running both UIs (the full UI when parked, the driving UI on the standard home screen).
